File: jpp/lab4/python/doubleWrapper.py
from polynomialType import PolynomialType


class DoubleWrapper(PolynomialType):
    value: float

    def __init__(self, number: float):
        self.value = number

    def neutral(self) -> "DoubleWrapper":
        return DoubleWrapper(0.0)

    def __add__(self, other: "DoubleWrapper") -> "DoubleWrapper":
        return DoubleWrapper(self.value + other.value)

    def __sub__(self, other: "DoubleWrapper") -> "DoubleWrapper":
        return DoubleWrapper(self.value - other.value)

    def __mul__(self, other: "DoubleWrapper") -> "DoubleWrapper":
        return DoubleWrapper(self.value * other.value)

    def __truediv__(self, other: "DoubleWrapper") -> "DoubleWrapper":
        return DoubleWrapper(self.value / other.value)

    def __pow__(self, other: int) -> "DoubleWrapper":
        return DoubleWrapper(pow(self.value, other))

    def __gt__(self, other: "DoubleWrapper") -> bool:
        return self.value > other.value

    def __lt__(self, other: "DoubleWrapper") -> bool:
        return self.value < other.value

    def __eq__(self, other: "DoubleWrapper") -> bool:
        return self.value == other.value

    def __str__(self) -> str:
        return str(self.value)

File: jpp/lab4/javaMvn/src/main/java/wilk/marcin/polynomial/DoubleWrapper.java
package wilk.marcin.polynomial;

public class DoubleWrapper implements PolynomialType<DoubleWrapper> {
  private double value;

  public DoubleWrapper(double value) {
    this.value = value;
  }

  @Override
  public DoubleWrapper neutral() {
    return new DoubleWrapper(0.0);
  }

  @Override
  public DoubleWrapper add(DoubleWrapper p) {
    return new DoubleWrapper(this.value + p.value);
  }

  @Override
  public DoubleWrapper sub(DoubleWrapper p) {
    return new DoubleWrapper(this.value - p.value);
  }

  @Override
  public DoubleWrapper prod(DoubleWrapper p) {
    return new DoubleWrapper(this.value * p.value);
  }

  @Override
  public DoubleWrapper div(DoubleWrapper p) {
    return new DoubleWrapper(this.value / p.value);
  }

  @Override
  public DoubleWrapper pow(long p) {
    return new DoubleWrapper(Math.pow(this.value, p));
  }

  @Override
  public boolean gt(DoubleWrapper other) {
    return this.value > other.value;
  }

  @Override
  public boolean lt(DoubleWrapper other) {
    return this.value < other.value;
  }

  @Override
  public boolean equals(Object obj) {
    if (this == obj) {
      return true;
    }

    if (obj == null) {
      return false;
    }

    if (getClass() != obj.getClass()) {
      return false;
    }

    DoubleWrapper other = (DoubleWrapper) obj;

    return this.value == other.value;
  }

  @Override
  public String toString() {
    return String.valueOf(this.value);
  }
}
